lib/postgrex/stream.ex

```diff
@@ -0,0 +1,65 @@
+defmodule Postgrex.Stream do
+  defstruct [:conn, :options, :params, :portal, :query, :state, :result, max_rows: 500 ]
+end
+
+defimpl Enumerable, for: Postgrex.Stream do
+  def reduce( stream, acc, fun ) do
+    start = fn
+      -> maybe_generate_portal( stream )
+    end
+    Stream.resource( start, &next/1, &close/1 ).( acc, fun )
+  end
+
+  def member?( _, _ ) do
+    {:error, __MODULE__}
+  end
+
+  def count( _) do
+    {:error, __MODULE__}
+  end
+
+  defp next( %Postgrex.Stream{ state: :done } = stream ) do
+    { :halt, stream }
+  end
+  defp next( %Postgrex.Stream{ conn: conn, params: params, options: options } = stream ) do
+    %Postgrex.Stream{ result: result } = stream = Postgrex.execute!( conn, stream, params, options )
+    emit_or_next( stream, result )
+  end
+  defp close( %Postgrex.Stream{ conn: conn } = stream ) do
+    DBConnection.close( conn, stream )
+  end
+
+  defp emit_or_next( stream, %Postgrex.Result{ rows: [] } ), do: next( stream )
+  defp emit_or_next( stream, result ),                       do: { [ result ], stream }
+
+  defp maybe_generate_portal( %Postgrex.Stream{ portal: nil } = stream ),
+    do: %Postgrex.Stream{ stream | portal: :erlang.ref_to_list( make_ref ) }
```

Comment:
I think we should always generated the portal otherwise this could lead to problems.